packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,301 @@
+					if (e instanceof APIError && e.isGatewayError()) {
+						// Gateway problem, wait for some additional time and set concurrency to 1
+						queue.concurrency = 1;
+						await new Promise((resolvePromise) =>
+							setTimeout(resolvePromise, Math.pow(2, gatewayErrors) * 5000)
+						);
+						gatewayErrors++;
+						// only count as a failed attempt after a few initial gateway errors
+						if (gatewayErrors >= MAX_UPLOAD_GATEWAY_ERRORS) {
+							attempts++;
+						}
+					} else {
+						attempts++;
+					}
+					return doUpload();
+				} else if (isJwtExpired(initializeAssetsResponse.jwt)) {
+					throw new FatalError(
+						`Asset upload took too long on bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}. Please try again.`
+					);
+				} else {
+					throw e;
+				}
+			}
+		};
+		// add to queue and run it if we haven't reached concurrency limit
+		void queue.add(() =>
+			doUpload().then((res) => {
+				bucketUploadCount++;
+				completionJwt = res.jwt || completionJwt;
+			})
+		);
+	}
+	queue.on("error", (error) => {
+		logger.error(error.message);
+		throw error;
+	});
+	await queue.onIdle();
+
+	// if queue finishes without receiving JWT from asset upload service (AUS)
+	// AUS only returns this in the final bucket upload response
+	if (!completionJwt) {
+		throw new FatalError(
+			"Failed to receive completion signal for file upload. Please try again.",
+			1
+		);
+	} else if (bucketUploadCount !== initializeAssetsResponse.buckets.length) {
+		throw new FatalError(
+			"Upload completion signal received unexpectedly early - we cannot confirm if all files were successfully uploaded. Please try again.",
+			1
+		);
+	}
+
+	const uploadMs = Date.now() - start;
+	const skipped = Object.keys(manifest).length - numberFilesToUpload;
+	const skippedMessage = skipped > 0 ? `(${skipped} already uploaded) ` : "";
+
+	logger.log(
+		`✨ Success! Uploaded ${numberFilesToUpload} file(s) ${skippedMessage}${formatTime(uploadMs)}\n`
+	);
+
+	return completionJwt;
+};
+
+// modified from /pages/validate.tsx
+const walk = async (
+	dir: string,
+	manifest: AssetManifest,
+	startingDir: string = dir
+) => {
+	const files = await readdir(dir);
+
+	let counter = 0;
+	await Promise.all(
+		files.map(async (file) => {
+			const filepath = path.join(dir, file);
+			const relativeFilepath = path.relative(startingDir, filepath);
+			const filestat = await stat(filepath);
+
+			if (filestat.isSymbolicLink()) {
+				return;
+			}
+
+			if (filestat.isDirectory()) {
+				manifest = await walk(filepath, manifest, startingDir);
+			} else {
+				if (counter >= MAX_ASSET_COUNT) {
+					throw new UserError(
+						`Max asset count exceeded: ${counter.toLocaleString()} files found.
+						You cannot have more than ${MAX_ASSET_COUNT.toLocaleString()} files in a deployment.
+						Ensure you have specified your build output directory correctly (currently set as ${startingDir}).`
+					);
+				}
+
+				const name = urlSafe(relativeFilepath);
```

Comment:
ah ha! This kinda already starts to touch on [my earlier comment](https://github.com/cloudflare/workers-sdk/pull/6374/files#r1705794267). Let's handle this in a chat with Carmen/me/you and figure out exactly what we want to do here.

Reply:
have we figured this out? and if not, can this PR be merged without having those details in place?

Reply:
I think it should be okay to leave for now, and I've added a [ticket](https://jira.cfdata.org/browse/WC-2455) to keep track.